Crash generating a declaration for invalid mapped type

Reported in https://github.com/Microsoft/TypeScript/pull/21316#issuecomment-366797259 

```sh
c:\test>type a.ts
type T0<T> = ({[K in keyof T]}) extends ({[key in K]: T[K]}) ? number : never;
c:\test>tsc --v
Version 2.8.0-dev.20180220

c:\test>tsc a.ts
a.ts(1,51): error TS2304: Cannot find name 'K'.
a.ts(1,57): error TS2304: Cannot find name 'K'.

c:\test>tsc -d c:\test\a.ts
C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:60550
                throw e;
                ^

TypeError: Cannot read property 'kind' of undefined
    at emitType (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55341:26)
    at emitMappedType (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55508:17)
    at emitType (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55380:28)
    at emitParenType (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55472:17)
    at emitType (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55374:28)
    at emitConditionalType (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55455:17)
    at emitType (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55370:28)
    at emitTypeWithNewGetSymbolAccessibilityDiagnostic (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55338:13)
    at writeTypeAliasDeclaration (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55839:13)
    at writeModuleElement (C:\Users\mhegazy\AppData\Roaming\npm\node_modules\typescript\lib\tsc.js:55631:28)
```

--- a/src/compiler/declarationEmitter.ts
+++ b/src/compiler/declarationEmitter.ts
@@ -608,7 +608,12 @@ namespace ts {
                         "?");
                 }
                 write(": ");
-                emitType(node.type);
+                if (node.type) {
+                    emitType(node.type);
+                }
+                else {
+                    write("any");
+                }
                 write(";");
                 writeLine();
                 decreaseIndent();

--- a/tests/baselines/reference/mappedTypeNoTypeNoCrash.errors.txt
+++ b/tests/baselines/reference/mappedTypeNoTypeNoCrash.errors.txt
@@ -1,0 +1,16 @@
+tests/cases/compiler/mappedTypeNoTypeNoCrash.ts(1,51): error TS2304: Cannot find name 'K'.
+tests/cases/compiler/mappedTypeNoTypeNoCrash.ts(1,51): error TS4081: Exported type alias 'T0' has or is using private name 'K'.
+tests/cases/compiler/mappedTypeNoTypeNoCrash.ts(1,57): error TS2304: Cannot find name 'K'.
+tests/cases/compiler/mappedTypeNoTypeNoCrash.ts(1,57): error TS4081: Exported type alias 'T0' has or is using private name 'K'.
+
+
+==== tests/cases/compiler/mappedTypeNoTypeNoCrash.ts (4 errors) ====
+    type T0<T> = ({[K in keyof T]}) extends ({[key in K]: T[K]}) ? number : never;
+                                                      ~
+!!! error TS2304: Cannot find name 'K'.
+                                                      ~
+!!! error TS4081: Exported type alias 'T0' has or is using private name 'K'.
+                                                            ~
+!!! error TS2304: Cannot find name 'K'.
+                                                            ~
+!!! error TS4081: Exported type alias 'T0' has or is using private name 'K'.

--- a/tests/baselines/reference/mappedTypeNoTypeNoCrash.js
+++ b/tests/baselines/reference/mappedTypeNoTypeNoCrash.js
@@ -1,0 +1,4 @@
+//// [mappedTypeNoTypeNoCrash.ts]
+type T0<T> = ({[K in keyof T]}) extends ({[key in K]: T[K]}) ? number : never;
+
+//// [mappedTypeNoTypeNoCrash.js]

--- a/tests/baselines/reference/mappedTypeNoTypeNoCrash.symbols
+++ b/tests/baselines/reference/mappedTypeNoTypeNoCrash.symbols
@@ -1,0 +1,9 @@
+=== tests/cases/compiler/mappedTypeNoTypeNoCrash.ts ===
+type T0<T> = ({[K in keyof T]}) extends ({[key in K]: T[K]}) ? number : never;
+>T0 : Symbol(T0, Decl(mappedTypeNoTypeNoCrash.ts, 0, 0))
+>T : Symbol(T, Decl(mappedTypeNoTypeNoCrash.ts, 0, 8))
+>K : Symbol(K, Decl(mappedTypeNoTypeNoCrash.ts, 0, 16))
+>T : Symbol(T, Decl(mappedTypeNoTypeNoCrash.ts, 0, 8))
+>key : Symbol(key, Decl(mappedTypeNoTypeNoCrash.ts, 0, 43))
+>T : Symbol(T, Decl(mappedTypeNoTypeNoCrash.ts, 0, 8))
+

--- a/tests/baselines/reference/mappedTypeNoTypeNoCrash.types
+++ b/tests/baselines/reference/mappedTypeNoTypeNoCrash.types
@@ -1,0 +1,11 @@
+=== tests/cases/compiler/mappedTypeNoTypeNoCrash.ts ===
+type T0<T> = ({[K in keyof T]}) extends ({[key in K]: T[K]}) ? number : never;
+>T0 : number
+>T : T
+>K : K
+>T : T
+>key : key
+>K : No type information available!
+>T : T
+>K : No type information available!
+

--- a/tests/cases/compiler/mappedTypeNoTypeNoCrash.ts
+++ b/tests/cases/compiler/mappedTypeNoTypeNoCrash.ts
@@ -1,0 +1,2 @@
+// @declaration: true
+type T0<T> = ({[K in keyof T]}) extends ({[key in K]: T[K]}) ? number : never;

